mlflow/types/schema.py

```diff
@@ -1329,3 +1330,128 @@ def __eq__(self, other) -> bool:
 
     def __repr__(self) -> str:
         return repr(self.params)
+
+
+def _map_field_type(field):
+    field_type_mapping = {
+        bool: "boolean",
+        int: "integer",
+        builtins.float: "float",
+        str: "string",
+        bytes: "binary",
+        dt.date: "datetime",
+    }
+    return field_type_mapping.get(field)
+
+
+def convert_dataclass_to_schema(dataclass):
+    """
+    Create a Schema object from a dataclass. The dataclass should have type hints
+    for each field. Fields in the dataclass can be composed of basic types or other dataclasses.
+    Only the top-level fields become ColSpecs directly. Lower-level fields are converted into
+    nested Object types.
+    """
+
+    inputs = []
+
+    for field_name, field_type in dataclass.__annotations__.items():
+        # Determine the type and handle Optional and List correctly
+        is_optional = False
+        effective_type = field_type
+
+        if get_origin(field_type) == Union:
+            if type(None) in get_args(field_type) and len(get_args(field_type)) == 2:
+                # This is an Optional type; determine the effective type excluding None
+                is_optional = True
+                effective_type = next(t for t in get_args(field_type) if t is not type(None))
+            else:
+                raise MlflowException(
+                    "Only Optional[...] is supported as a Union type in dataclass fields"
+                )
+
+        if get_origin(effective_type) == list:
+            # It's a list, check the type within the list
+            list_type = get_args(effective_type)[0]
+            if is_dataclass(list_type):
+                dtype = convert_dataclass_to_nested_object(list_type)  # Convert to nested Object
+                inputs.append(
+                    ColSpec(type=Array(dtype=dtype), name=field_name, required=not is_optional)
+                )
+            else:
```

Comment:
If list_type is not in the mapping this raises strange error. BTW let's be clear that we only support List[dataclass] or List[primitives] on the error message :D